validator/src/validator/updater.rs

```diff
@@ -0,0 +1,78 @@
+use std::process::Command;
+use std::thread::{self, sleep};
+use std::time::Duration;
+use tracing::{info, error};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum UpdaterError {
+    #[error("Git pull failed: {0}")]
+    PullFailed(String),
+    
+    #[error("Build failed: {0}")]
+    BuildFailed(String),
+    
+    #[error(transparent)]
+    IoError(#[from] std::io::Error),
+    
+    #[error(transparent)]
+    Utf8Error(#[from] std::string::FromUtf8Error)
+}
+
+pub struct Updater {
+    check_interval: Duration,
+    git_branch: String,
+}
+
+impl Updater {
+    pub fn new(check_interval: Duration, git_branch: String) -> Self {
+        Self { 
+            check_interval,
+            git_branch,
+        }
+    }
+
+    pub fn spawn(self) -> thread::JoinHandle<()> {
+        thread::spawn(move || {
+            loop {
+                if let Err(e) = self.check_and_update() {
+                    error!("Update check failed: {}", e);
+                }
+                sleep(self.check_interval);
+            }
+        })
+    }
+
+    fn check_and_update(&self) -> Result<(), UpdaterError> {
+        let pull = Command::new("git")
+            .args(["pull", "--ff-only", "origin", &self.git_branch])
```

Comment:
I don't think you should specify the branch. Can instead just not include `origin` and it'd use the current one